### ISSUE TITLE
findUsers on large AD using pagination

### ISSUE THREAD

**Issue by avallete** [jsumners/node-activedirectory#20]
Hi !

I'm face some troubles to use the `findUsers` function on large repositories.

I want to retrieve the users using the pagination system. So I have add

```
pageSize: 10,
paged: true
```

Into my AD instance options.

But when I try to retrieve users like:
```
 ad.findUsers((err, users) => {
    console.log(users.length);
    if (err) {
      console.log('error occured');
    }
    if (!users) {
      console.log('no users');
    }
    else {
      console.log(users)
    }
});
```

I see the pageSize argument is working. But I don't understand how to iterate on next pages to retrieve the next 10 users, and so on.

If anyone have an pseudo-code example to do this kind of things it would be greatly appreciated :).

**Comment by jsumners** [comment]
Are you not seeing _all_ of the found users in the callback result? The module is supposed to be handling the paging for you, if I remember correctly.

**Comment by avallete** [comment]
Nop, that was what I thinked too, but into the `callback` function I've only the X first users from my AD (X matching the pageSize parameter).

**Comment by jsumners** [comment]
I'm really not sure. I'd pass in an instance of [pino](https://github.com/pinojs/pino) with the level set to "trace" and see what sort of information gets logged.

**Comment by jurjendijkstra** [comment]
It took me some time to experiment and found that "paged:true" should not go on the AD instance options, but on the findUsers call :

`return ad.findUsers({baseDN: settings.baseDN, paged: true}).then( ... )`

we have about 3500 matching users and this returns them all in one single call

**Comment by jsumners** [comment]
@jurjendijkstra nice work. I think the documentation needs some clarification.

**Comment by avallete** [comment]
Thank's all for your help, as @jurjendijkstra says, it work when you put the `{paged: true}` as findUsers arguments.

```
ad.findUsers({paged: true}, (err, users) => { console.log(users.length) })
```

Give me my +1500 users :)

**Comment by jsumners** [comment]
@xjtumalin as the other messages in this thread say: you will get back an array of all results.

**Comment by xjtumalin** [comment]
@jsumners thank you for replying so quick. I'm wondering if there's a better way to do it. It would be very heavy workload for the nodejs server and AD server if there're plenty of users in the AD.

**Comment by jsumners** [comment]
You are right. It is most definitely not recommended to issue a query that would return a significant number of results. But that's what the original library did, and this one is just a refactor of it.

The way around it would be to expose the paging event so that users can handle it themselves. PRs are welcome.

**Comment by jaypark96** [comment]
Hi, I'm currently working with AD and using {paged: true} within the findUsers function, it only returns a max of 5000 users for me. I'm expecting the result to be about > 100,000 users though. I tried working around it and just doing getUsersForGroup however the number of groups caps out at 5000 as well.

**Comment by jsumners** [comment]
I cannot devote any resources to fixing this. I no longer have access to an Active Directory instance, and I just don't have the time.

I am very willing to get a fix merged if a PR is submitted. And I'm open to adding maintainers if anyone is up for it.

**Comment by jurjendijkstra** [comment]
It is an LDAP policy to return no more than 5000 members, so it's not the fault of this node package. See for example  https://support.microsoft.com/en-us/help/2009267/windows-server-2008-and-newer-domain-controller-returns-only-5000-valu

**Comment by jaypark96** [comment]
@jurjendijkstra do you happen to know a fix around this?

**Comment by jurjendijkstra** [comment]
no sorry, I do not know. I have never needed such large numbers. If it is even possible then it should be somewhere in the ldapjs module. If you are lucky and the results are sorted (I dont know if they are) then perhaps you may use the ldapjs GreaterThanEqualsFilter to get members whose cn is greater than the greatest cn in the batch you just got?
